src/ucp/rndv/proto_rndv.inl

```diff
@@ -265,6 +270,51 @@ ucp_proto_rndv_bulk_max_payload(ucp_request_t *req,
     return max_payload;
 }
 
+/**
+ * Return the fragment size to ensure alignment for buffer of main data
+ * portion. Either return max_payload when the buffer is initially aligned,
+ * or alignment fragment already sent.
+ */
+static UCS_F_ALWAYS_INLINE size_t
+ucp_proto_rndv_adjust_to_alignment(ucp_request_t *req,
+                                   const ucp_proto_rndv_bulk_priv_t *rpriv,
+                                   const ucp_proto_multi_lane_priv_t *lpriv,
+                                   const size_t max_payload)
+{
+    size_t min_frag     = rpriv->mpriv.min_frag;
+    size_t total_offset = ucp_proto_rndv_request_total_offset(req);
+    size_t total_length = ucp_proto_rndv_request_total_length(req);
+    size_t align_thresh, align_size;
+    void *buffer;
+    unsigned buffer_padding;
+    size_t align;
+    uct_iface_attr_t *attrs;
+
+    ucs_assertv(req->send.state.dt_iter.dt_class == UCP_DATATYPE_CONTIG,
+                "dt_class=%d (%s)", req->send.state.dt_iter.dt_class,
+                ucp_datatype_class_names[req->send.state.dt_iter.dt_class]);
+
+    align_thresh = req->send.ep->worker->context->config.ext.rndv_align_thresh;
+    if (align_thresh > total_length) {
+        return max_payload;
+    }
+
+    attrs = ucp_worker_iface_get_attr(req->send.ep->worker,
+                                      ucp_ep_get_rsc_index(req->send.ep,
+                                                           lpriv->super.lane));
+    align  = attrs->cap.get.opt_zcopy_align;
+    buffer = UCS_PTR_BYTE_OFFSET(req->send.state.dt_iter.type.contig.buffer,
+                                 total_offset);
+    buffer_padding = ((size_t)buffer) % align;
+    align_size     = ucs_align_up((min_frag + buffer_padding),
+                                  align) - buffer_padding;
+    if ((!buffer_padding) || (align_size >= max_payload)) {
+        return max_payload;
+    }
+    req->flags |= UCP_REQUEST_FLAG_RNDV_ALIGNING_FRAG;
```

Comment:
IMO adding request flag is pretty intrusive; instead, can we return an indication from send callback to progress_multi that multi_lane_idx should not be advanced?